pkg/gadgets/dns/tracer/tracer.go

```diff
@@ -0,0 +1,226 @@
+		if err != nil {
+			return -1, err
+		}
+		defer netnsHandle.Close()
+		err = netns.Set(netnsHandle)
+		if err != nil {
+			return -1, err
+		}
+
+		// Switch back to the original namespace
+		defer netns.Set(origns)
+	}
+
+	sock, err := syscall.Socket(syscall.AF_PACKET, syscall.SOCK_RAW|syscall.SOCK_NONBLOCK|syscall.SOCK_CLOEXEC, int(htons(syscall.ETH_P_ALL)))
+	if err != nil {
+		return -1, err
+	}
+	sll := syscall.SockaddrLinklayer{
+		Ifindex:  0, // 0 matches any interface
+		Protocol: htons(syscall.ETH_P_ALL),
+	}
+	if err := syscall.Bind(sock, &sll); err != nil {
+		return -1, err
+	}
+	return sock, nil
+}
+
+// htons converts the unsigned short integer hostshort from host byte order to network byte order.
+func htons(i uint16) uint16 {
+	b := make([]byte, 2)
+	binary.BigEndian.PutUint16(b, i)
+	return *(*uint16)(unsafe.Pointer(&b[0]))
+}
+
+func NewTracer() (*Tracer, error) {
+	spec, err := ebpf.LoadCollectionSpecFromReader(bytes.NewReader(ebpfProg))
+	if err != nil {
+		return nil, fmt.Errorf("failed to load asset: %s", err)
+	}
+
+	t := &Tracer{
+		spec:        spec,
+		attachments: make(map[string]*link),
+	}
+
+	return t, nil
+}
+
+func (t *Tracer) Attach(key string, pid uint32, f func(name, pktType string)) error {
+	if _, ok := t.attachments[key]; ok {
+		return fmt.Errorf("key already attached: %q", key)
+	}
+
+	coll, err := ebpf.NewCollectionWithOptions(t.spec, ebpf.CollectionOptions{Programs: ebpf.ProgramOptions{LogSize: ebpf.DefaultVerifierLogSize * 100}})
+	if err != nil {
+		return fmt.Errorf("failed to create BPF collection: %s", err)
+	}
+
+	rd, err := perf.NewReader(coll.Maps[BPF_MAP_NAME], os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("failed to get a perf reader: %s", err)
+	}
+
+	prog, ok := coll.Programs[BPF_PROG_NAME]
+	if !ok {
+		return fmt.Errorf("Failed to find BPF program %q", BPF_PROG_NAME)
+	}
+
+	sockFd, err := openRawSock(pid)
+	if err != nil {
+		return fmt.Errorf("Failed to open raw socket: %s", err)
+	}
+
+	if err := syscall.SetsockoptInt(sockFd, syscall.SOL_SOCKET, SO_ATTACH_BPF, prog.FD()); err != nil {
+		return fmt.Errorf("Failed to attach BPF program: %s", err)
+	}
+
+	l := &link{
+		collection: coll,
+		sockFd:     sockFd,
+		perfRd:     rd,
+		stop:       make(chan struct{}),
+	}
+	t.attachments[key] = l
+
+	go t.listen(rd, f)
+
+	return nil
+}
+
+func parseDNSEvent(rawSample []byte) (ret string, pktType string) {
+	// Parse the perf event entry into an Event structure.
+	name := make([]byte, C.MAX_DNS_NAME)
+	copy(name, rawSample)
+	var dummy C.struct_event_t
+	if len(rawSample) >= int(unsafe.Offsetof(dummy.pkt_type))+1 {
+		pktType = fmt.Sprintf("%d", rawSample[unsafe.Offsetof(dummy.pkt_type)])
+	}
+
+	// Convert name into a string with dots
+	for i := 0; i < C.MAX_DNS_NAME; i++ {
+		length := int(name[i])
+		if length == 0 {
+			break
+		}
+		if i+1+length < C.MAX_DNS_NAME {
+			ret += string(name[i+1:i+1+length]) + "."
+		}
+		i += length
+	}
```

Comment:
Could this work? 

```
	// Convert name into a string with dots
	i := 0
	while i < C.MAX_DNS_NAME {
		length := int(name[i])
		if length == 0 {
			break // last one, we're done 
		}
		name[i] = '.'
		i += length
	}
```

Reply:
I think that would generate `.www.kinvolk.io` instead of `www.kinvolk.io.`